File: insight/backend/app/models/submission.py
"""Submission models."""

import enum
from datetime import date, datetime

from sqlalchemy import (
    Column,
    String,
    Text,
    Boolean,
    Date,
    DateTime,
    Numeric,
    Enum,
    ForeignKey,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.models.base import BaseModel


class SubmissionStatus(str, enum.Enum):
    """Submission statuses."""

    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    AUTO_SUBMITTED = "auto_submitted"


class Submission(BaseModel):
    """Submission model."""

    __tablename__ = "submissions"

    form_id = Column(UUID(as_uuid=True), ForeignKey("forms.id"), nullable=False)
    submitted_by = Column(UUID(as_uuid=True), ForeignKey("team.id"), nullable=False)
    submission_date = Column(Date, nullable=False, default=date.today)
    submission_time = Column(DateTime(timezone=True), default=datetime.now)
    status = Column(Enum(SubmissionStatus), default=SubmissionStatus.IN_PROGRESS)
    completion_percentage = Column(Numeric(5, 2), default=0.0)
    is_auto_submitted = Column(Boolean, default=False)

    # Relationships
    form = relationship("Form", back_populates="submissions")
    answers = relationship(
        "SubmissionAnswer", back_populates="submission", cascade="all, delete-orphan"
    )


class SubmissionAnswer(BaseModel):
    """Submission answer model."""

    __tablename__ = "submission_answers"
    __table_args__ = (UniqueConstraint("submission_id", "field_id"),)

    submission_id = Column(
        UUID(as_uuid=True),
        ForeignKey("submissions.id", ondelete="CASCADE"),
        nullable=False,
    )
    field_id = Column(UUID(as_uuid=True), ForeignKey("fields.id"), nullable=False)
    answer_value = Column(Text)
    file_url = Column(String)
    answered_at = Column(DateTime(timezone=True), default=datetime.now)

    # Relationships
    submission = relationship("Submission", back_populates="answers")

File: insight/packages/insight_core/lib/src/models/submission.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'submission.freezed.dart';
part 'submission.g.dart';

enum SubmissionStatus {
  @JsonValue('in_progress')
  inProgress,
  @JsonValue('completed')
  completed,
  @JsonValue('auto_submitted')
  autoSubmitted,
}

@freezed
class Submission with _$Submission {
  const factory Submission({
    required String id,
    required String formId,
    required String submittedBy,
    required DateTime submissionDate,
    required DateTime submissionTime,
    @Default(SubmissionStatus.inProgress) SubmissionStatus status,
    @Default(0.0) double completionPercentage,
    @Default(false) bool isAutoSubmitted,
    @Default([]) List<SubmissionAnswer> answers,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Submission;

  factory Submission.fromJson(Map<String, dynamic> json) =>
      _$SubmissionFromJson(json);
}

@freezed
class SubmissionAnswer with _$SubmissionAnswer {
  const factory SubmissionAnswer({
    required String id,
    required String submissionId,
    required String fieldId,
    required String fieldLabel,
    String? answerValue,
    String? value,
    String? fileUrl,
    required DateTime answeredAt,
  }) = _SubmissionAnswer;

  factory SubmissionAnswer.fromJson(Map<String, dynamic> json) =>
      _$SubmissionAnswerFromJson(json);
}
